Created two new files: joint_degree_seq.py and test_joint_degree_seq.…

mgjoka opened the pull request:
I created two new files: joint_degree_seq.py and test_joint_degree_seq.py . Inside joint_degree_seq.py I added functions for validity check (is_valid_joint_degree) and graph construction (joint_degree_model) for a given joint degree. Inside  test_joint_degree_seq.py I added tests for the latter two main functions. File joint_degree_seq.py also has a helper function named _neighbor_switch (which is called inside joint_degree_model).

This is a follow up for feature proposed [by me at this post](https://github.com/networkx/networkx/issues/1996)  and accepted by Aric Hagberg.  Let me know if you have any feedback.

MridulS commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,283 @@
+"""Generate graphs with a given joint degree 
+"""
+#    BSD license.
```
The top of the file should look something like this
```
#    Copyright (C) 2004-2016 by
#    Aric Hagberg <hagberg@lanl.gov>
#    Dan Schult <dschult@colgate.edu>
#    Pieter Swart <swart@lanl.gov>
#    All rights reserved.
#    BSD license.
#
# Authors:  Minas Gjoka (minas.gjoka@gmail.com)
"""Generate graphs with a given joint degree """
```

hagberg replied:
No - it doesn't need to look like that.  We do not ask the copyright be assigned.  The code should be licensed by the author with BSD or other compatible license in order to be included.

dschult replied:
So, for example, you could make the top of the file look like:
```
#    Copyright (C) 2016 by
#    Minas Gjoka
#    All rights reserved.
#    BSD license.
#
# Author:  Minas Gjoka (minas.gjoka@gmail.com)
"""Generate graphs with a given joint degree """
import random
```

There is a description of file-header format in the NetworkX Github Wiki.
The decision on copyright and license is yours as described by @hagberg above.

MridulS commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,283 @@
+
+import networkx as nx
+
+__author__ = "\n".join(['Minas Gjoka <minas.gjoka@gmail.com>'])
```
This is not required, author information is preserved in the git commit and the file header above.

MridulS commented on networkx/generators/tests/test_joint_degree_seq.py:
```diff
@@ -0,0 +1,57 @@
+        nkk_G = degree_mixing_dict(G,normalized=False)
+        
+        #assert that the given joint degree is equal to the generated graph's joint degree
+        assert_true(nkk_g == nkk_G)
```
new line

MridulS commented:
You need to add `from networkx.generators.joint_degree_seq import *` to  `networkx/generators/__init.py`.

MridulS commented:
You also need to add this
```
Joint Degree Sequence
--------------------
.. automodule:: networkx.generators.joint_degree_seq
.. autosummary::
   :toctree: generated/

   is_valid_joint_degree
   joint_degree_model
```
to `doc/source/reference/generators.rst`

MridulS commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,283 @@
+    >>> joint_degree_dict = {1: {4: 1},
+                             2: {2: 2, 3: 2, 4: 2},
+                             3: {2: 2, 4: 1},
+                             4: {1: 1, 2: 2, 3: 1}}
```
This should be
```
>>> joint_degree_dict = {1: {4: 1},
...                      2: {2: 2, 3: 2, 4: 2},
...                      3: {2: 2, 4: 1},
...                      4: {1: 1, 2: 2, 3: 1}}
```

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+    nk = {}
+    for k in nkk:
+        if k>0:
+            k_size = float( sum(val for val in nkk[k].itervalues()) ) / float( k )
```
This would be simpler as follows. First, add `from __future__ import division` to the head of the file (after the module docstring), then do:

    k_size = sum(nkk[k].values()) / k

mgjoka replied:
just to confirm: would this return a float ?

also, is it not faster and better in terms of  memory consumption  to use nkk[k].itervalues()  ?

mgjoka replied:
turns out that  nkk[k].itervalues() is the reason for python 3.x failures. alright, I will use the suggested way.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+    for v in G[w]: 
+        if (v not in w_prime_neighbs) and (v!=w_prime):
+            t = v
+            break;
```
No semicolons necessary in python :)

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+        random.seed(seed)
+
+    # compute degree vector from joint degree nkk
+    nk = { deg:(sum(val for val in nkk[deg].itervalues())/deg) for deg in nkk if deg>0 }
```
This should maybe be

    nk = {source: sum(target.values()) // source for source, target in nkk.items() if source > 0}

but with better names than `source` and `target`. Notice the integer division, since you are guaranteed integers by invoking `is_valid_joint_degree()` above.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+    nk = { deg:(sum(val for val in nkk[deg].itervalues())/deg) for deg in nkk if deg>0 }
+    
+    # start with empty N-node graph
+    N = int(sum(deg for deg in nk.itervalues()))
```
Again, just `n = sum(nk.values())`.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+    
+    # populate h_degree_nodelist and h_node_residual
+    nodeid = 0    
+    for degree, numNodes in nk.iteritems(): 
```
Need to use `.items()` instead of `.iteritems()` everywhere.

jfinkels replied:
Also, no camel case variable names in Python except for class names.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+    # populate h_degree_nodelist and h_node_residual
+    nodeid = 0    
+    for degree, numNodes in nk.iteritems(): 
+        h_degree_nodelist[degree] = range(nodeid, nodeid+int(numNodes))         
```
Probably don't need `int` here.

mgjoka commented:
I see that the Travis build failed on the python 3.x versions. Actually, I have  only tested the algorithms on all python 2.x versions. I will go ahead and see what is the problem on the 3.x versions.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+                    l_unsat = set(w for w in l_nodes if h_node_residual[w]>0) 
+                else:
+                    l_unsat = k_unsat
+                    n_edges_add = nkk[k][l]/2
```
Is this guaranteed to be an even number? If this is definitely going to be integer division, use `//` instead of `/`.

jfinkels commented on networkx/generators/joint_degree_seq.py:
```diff
@@ -0,0 +1,286 @@
+                while n_edges_add > 0:
+                    
+                    # randomly pick nodes v and w that have degrees k and l
+                    v = k_nodes[ random.randint(0,k_size-1) ]
```
Use `random.randrange(k_size)` here, it matches Python's ubiquitous `range` idiom more closely.

jfinkels commented:
All of your docstrings need to be wrapped to 79 characters per line. Also, they need to be formatted so that they are rendered well when compiled to HTML by Sphinx: for example, function arguments should be wrapped by asterisks, and code snippets should be surrounded by backticks. See http://www.sphinx-doc.org/en/stable/ for more information on formatting docstrings.

mgjoka commented:
I addressed all comments from the second review. 

Regarding the docstrings, I made all lines be less than 80 characters. I also added asterisks around function arguments, in most places where they occur. It would help if you could show me some examples of source files in networkx that strictly adhere to Sphinx requirements.

mgjoka commented:
I made another push. No changes to the source code itself. Just removed unnecessary whitespace (end of lines) and added necessary whitespace e.g. around equality/inequality conditions in "if statements", function parameters etc.

mgjoka commented:
It seems that my last commit has passed all checks. Are there any updates on the merging process? Perhaps a timeline on when the commit will be merged ?

jfinkels commented on doc/source/reference/generators.rst:
```diff
@@ -240,3 +240,13 @@ Triads
+
+
+Joint Degree Sequence
+--------------------
```
Number of dashes needs to match number of characters on previous line.

jfinkels commented on networkx/generators/__init__.py:
```diff
@@ -19,3 +19,4 @@
 from networkx.generators.social import *
 from networkx.generators.stochastic import *
 from networkx.generators.triads import *
+from networkx.generators.joint_degree_seq import *
```
This should be in alphabetical order.

mgjoka commented:
Addressed all comments.

jfinkels commented:
Here's an example of a better docstring for `is_valid_joint_degree`, with inline explanations of why I make these suggestions. See also [PEP 0257][1] for docstring conventions:

    def is_valid_joint_degree(joint_degrees):

Change the argument name to be meaningful.

        """Checks whether the given joint degree dictionary is realizable as
        a simple graph.

Explain what the function does without describing the implementation details. Use four-space indentation after the first line.

        A *joint degree dictionary* is a dictionary of dictionaries, in
        which entry ``joint_degrees[k][l]`` is an integer representing the
        number of edges joining nodes of degree *k* with nodes of degree
        *l*. Such a dictionary is realizable as a simple graph if and only
        if the following conditions are satisfied.

Explain what a "joint degree dictionary" is in as close to plain English as possible while still maintaining mathematical readability.

        - each entry must be an integer,
        - the total number of nodes of degree *k*, computed by
          ``sum(joint_degrees[k].values()) / k``, must be an integer,
        - the total number of edges joining nodes of degree *k* with nodes
          of degree *l* cannot exceed the total number of possible edges,
        - each diagonal entry ``joint_degrees[k][k]`` must be even (this is
          a convention assumed by the :func:`joint_degree_model` function).

This is an unordered list of conditions, the numbers are not necessary. Code and variable names are correctly formatted using the ReStructured Text format. Decide whether the code snippets are there as a reminder to you as the implementer or as important information to the user; if the former, then remove it and place it in a comment in the code.

        Parameters
        ----------
        joint_degrees :  dictionary of dictionary of integers
            A joint degree dictionary in which entry ``joint_degrees[k][l]``
            is the number of edges joining nodes of degree *k* with nodes of
            degree *l*.

        Returns
        -------
        bool
            Whether the given joint degree dictionary is realizable as a
            simple graph.

        References
        ----------
        .. [1] M. Gjoka, M. Kurant, A. Markopoulou, "2.5K Graphs: from Sampling
           to Generation", IEEE Infocom, 2013.

        """

Similar changes can be made to the other functions as well.

[1]: https://www.python.org/dev/peps/pep-0257/

mgjoka commented:
Comments about docstrings and variable names addressed.

hagberg commented:
Thanks for making those formatting changes.  It looks good (though I didn't check the formatted doc output).  One more small suggestion would be to rename the function to `joint_degree_graph` instead of `joint_degree_model`.  There are some cases where functions that return graphs aren't named that way (e.g. `configuration_model`) but I prefer the `graph` name approach.

mgjoka commented:
ok, I followed your suggestion and changed the function name to joint_degree_graph.

mgjoka commented:
Can someone re-run the Travis build? The error occurring is not because of the add-joint-degree branch but due to an unrelated error inside networkx.convert_matrix.to_pandas_dataframe which this branch never uses.

hagberg commented:
Any comment on this approach vs http://arxiv.org/abs/1503.06725?

mgjoka commented:
Regarding http://arxiv.org/abs/1503.06725  :

That is indeed very relevant work. The computation complexity of [1503.06725] is  O(M*N), whereas for the algorithm that I contributed it is O(M*max_degree). Since typically max_degree<<N the current algorithm is going to be much faster for larger graphs. However, the algorithm of [1503.06725] does more than produce an exact simple graph given a JDM. It also allows one to more efficiently sample from the  space of graphs with a given JDM. Therefore, I think that  despite the added computational complexity, because the algorithm of [1503.06725] brings some benefits it would be advisable to integrate it within networkx.   That is assuming someone is willing to do the work.

hagberg commented:
HI @mgjoka, thanks for that very helpful comment.   It would be a good project for someone to implement  that method too - but is a separate issue.  Thanks for your contribution here.
